geopandas/_vectorized.py

```diff
@@ -1095,10 +1095,21 @@ def bounds(data):
 #
 
 
-def transform(data, func):
+def transform(data, func, include_z=None):
     if compat.USE_SHAPELY_20:
-        coords = shapely.get_coordinates(data)
-        new_coords = func(coords[:, 0], coords[:, 1])
+        if include_z is None:
+            # if not specified explicitly, try to infer from first geometry
+            include_z = False
+            if len(data):
+                is_geometry = shapely.is_geometry(data)
+                if is_geometry.any():
+                    first_geom = data[np.argmax(is_geometry)]
+                    include_z = first_geom.has_z
```

Comment:
I am doing some effort here to infer it from the first value that is an actual geometry (in the idea to avoid a more costly computation to check it for all values). 
But we could also make this simpler and just do `shapely.has_z(data).any()` to see if any of the geometries has z values. 

That doesn't seem to be that much slower:

```
In [72]: data = shapely.points(np.random.randn(1_000_000, 2))

In [73]: %%timeit
    ...: is_geometry = shapely.is_geometry(data)
    ...: if is_geometry.any():
    ...:     first_geom = data[np.argmax(is_geometry)]
    ...:     include_z = first_geom.has_z
16.2 ms ± 810 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [74]: %%timeit
    ...: include_z = shapely.has_z(data).any()
    ...: 
    ...: 
21.1 ms ± 652 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Reply:
I still think it is a bad idea to assume that the whole array has or does not have Z coords. If you use `any`, you will get empty geometries.

Reply:
(yeah, to be clear posted the above before I read your comment ;))